File: rust/crates/crispy-core/src/traits/dvr_repository.rs
use crate::database::DbError;
use crate::models::{Recording, StorageBackend, TransferTask};

/// Persistence contract for DVR recordings, storage backends, and transfer tasks.
pub trait DvrRepository {
    fn save_recording(&self, rec: &Recording) -> Result<(), DbError>;
    fn load_recordings(&self) -> Result<Vec<Recording>, DbError>;
    fn update_recording(&self, rec: &Recording) -> Result<(), DbError>;
    fn delete_recording(&self, id: &str) -> Result<(), DbError>;
    fn save_storage_backend(&self, backend: &StorageBackend) -> Result<(), DbError>;
    fn load_storage_backends(&self) -> Result<Vec<StorageBackend>, DbError>;
    fn delete_storage_backend(&self, id: &str) -> Result<(), DbError>;
    fn save_transfer_task(&self, task: &TransferTask) -> Result<(), DbError>;
    fn load_transfer_tasks(&self) -> Result<Vec<TransferTask>, DbError>;
    fn update_transfer_task(&self, task: &TransferTask) -> Result<(), DbError>;
    fn delete_transfer_task(&self, id: &str) -> Result<(), DbError>;
}

File: app/flutter/lib/features/dvr/domain/repositories/dvr_repository.dart
import '../entities/recording.dart';
import '../entities/storage_backend.dart';
import '../entities/transfer_task.dart';
import '../entities/commercial_marker.dart';

/// Repository contract for DVR recording, storage backend, and
/// transfer task operations.
///
/// Implemented by the infrastructure layer backed by the Rust
/// crispy-core engine via CacheService.
abstract interface class DvrRepository {
  // ── Recordings ─────────────────────────────────────

  /// Load all recordings.
  Future<List<Recording>> loadRecordings();

  /// Save (insert or replace) a recording.
  Future<void> saveRecording(Recording recording);

  /// Update a recording's status and file info.
  Future<void> updateRecording(Recording recording);

  /// Delete a recording by [id].
  Future<void> deleteRecording(String id);

  /// Get commercial markers for [recordingId].
  Future<List<CommercialMarker>> getRecordingMarkers(String recordingId);

  // ── Storage Backends ───────────────────────────────

  /// Load all configured storage backends.
  Future<List<StorageBackend>> loadStorageBackends();

  /// Save (insert or replace) a storage backend.
  Future<void> saveStorageBackend(StorageBackend backend);

  /// Delete a storage backend by [id].
  Future<void> deleteStorageBackend(String id);

  // ── Transfer Tasks ─────────────────────────────────

  /// Load all transfer tasks.
  Future<List<TransferTask>> loadTransferTasks();

  /// Save (insert or replace) a transfer task.
  Future<void> saveTransferTask(TransferTask task);

  /// Update a transfer task's status and progress.
  Future<void> updateTransferTask(TransferTask task);

  /// Delete a transfer task by [id].
  Future<void> deleteTransferTask(String id);
}
